File: StrategyPattern/StrategyDemo.java
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;

interface DiscountStrategy {
  double calculateDiscount(double total);
}

class NoDiscountStrategy implements DiscountStrategy {

  @Override
  public double calculateDiscount(double total) {
    return 0;
  }

}

class StandardDiscountStrategy implements DiscountStrategy {
  @Override
  public double calculateDiscount(double total) {
    return total * 0.10;
  }
}

class HighValueDiscountStrategy implements DiscountStrategy {
  @Override
  public double calculateDiscount(double total) {

    // WE CAN CREATE A MAP IF THERE ARE TOO MANY IF ELSE
    if (total > 1000) {
      return total * .20;
    }

    return total * .10;

  }
}

class VipDiscountStrategy implements DiscountStrategy {
  @Override
  public double calculateDiscount(double total) {
    return total * .30;
  }
}

class Product {
  String name;
  double unitPrice;
  double quantity;

  Product(String name, double unitPrice, int quantity) {
    this.name = name;
    this.unitPrice = unitPrice;
    this.quantity = quantity;
  }
}

class Order {
  List<Product> productList = new ArrayList<>();
  // String discountStrategy = "no discount";
  DiscountStrategy strategy;

  Order(DiscountStrategy strategy) {
    this.strategy = strategy;
  }

  public void addProduct(Product product) {
    productList.add(product);
  }

  public List<Product> getProductList() {
    return productList;
  }

  // void setDiscountStrategy(String strategy) {
  // this.discountStrategy = strategy.toLowerCase();
  // }

  void checkOut() {
    double total = 0;
    double discount = 0;
    for (var product : productList) {
      total += product.unitPrice * product.quantity;
    }

    discount = this.strategy.calculateDiscount(total);

    System.out.println("Order details");
    for (var product : productList) {
      System.out.println(MessageFormat.format("{0} {1} {2} {3}", product.name, product.unitPrice, product.quantity,
          product.unitPrice * product.quantity));
    }
    System.out.println(MessageFormat.format("Amount: {0}", total));
    System.out.println(MessageFormat.format("Discount: {0}", discount));
    System.out.println(MessageFormat.format("Net: {0}", total - discount));
  }
}

public class StrategyDemo {

  public static void main(String[] args) {
    var order = new Order(new VipDiscountStrategy());
    order.addProduct(new Product("iPhone", 1_00_000, 1));
    order.addProduct(new Product("AirPods", 30_000, 2));
    order.addProduct(new Product("Macbook", 2_20_000, 1));

    // order.setDiscountStrategy("standard discount");

    order.checkOut();
  }
}

File: StrategyPattern/StrategyDemo.py
from abc import ABC, abstractmethod

class Product:
    def __init__(self, name: str, unit_price: float, quantity: int) -> None:
        self.name: str = name
        self.unit_price: float = unit_price
        self.quantity: int = quantity


class Order:
    def __init__(self) -> None:
        self.product_list: list[Product] = []
        self.discount_strategy: str = "no discount"

    def add_product(self, product: Product) -> None:
        self.product_list.append(product)

    def get_product_list(self) -> list[Product]:
        return self.product_list

    def set_discount_strategy(self, strategy: str) -> None:
        self.discount_strategy = strategy.lower()

    def check_out(self) -> None:
        total: float = 0
        discount: float = 0
        for product in self.product_list:
            total += product.unit_price * product.quantity

        if (self.discount_strategy == "no discount"):
            discount = 0
        elif (self.discount_strategy == "standard discount"):
            discount = total * .10
        elif (self.discount_strategy == "high value discount"):
            if (total > 1000):
                discount = total * .20
            else:
                discount = total * .10
        elif (self.discount_strategy == "vip discount"):
            discount = total * .30

        print("Order details")
        for product in self.product_list:
            print(
                f"{product.name} {product.unit_price} {product.quantity} {product.unit_price * product.quantity}")
        print(f"Amount: {total}")
        print(f"Discount: {discount}")
        print(f"Net: {total - discount}")


order = Order()
order.add_product(Product("iPhone", 1_00_000, 1))
order.add_product(Product("AirPods", 30_000, 2))
order.add_product(Product("Macbook", 2_20_000, 1))
order.set_discount_strategy("standard discount")
order.check_out()
